RickAndMorty-iOS/Views/Characters/CharactersViewController.swift

```diff
@@ -54,49 +58,38 @@ class CharactersViewController: UIViewController {
     }
 
     @objc func onRefresh() {
-        characterInfo = []
         viewModel.currentPage = 1
     }
 
     func loadMore() {
-        characterInfo = []
         viewModel.currentPage += 1
     }
 }
 
 // MARK: - UICollectionViewDataSource
-extension CharactersViewController: UICollectionViewDataSource {
-    func collectionView(_ collectionView: UICollectionView, numberOfItemsInSection section: Int) -> Int {
-        return characterInfo.count
-    }
-
-    func collectionView(_ collectionView: UICollectionView, cellForItemAt indexPath: IndexPath) -> UICollectionViewCell {
-        let cell = collectionView.dequeueReusableCell(withReuseIdentifier: CharactersGridCell.identifier,
-                                                            for: indexPath) as? CharactersGridCell
-
-        if !characterInfo.isEmpty {
-            let name = self.characterInfo[indexPath.item].name
-            guard let image = characterInfo[indexPath.item].image else { fatalError("Image not found") }
-            cell?.characterNameLabel.text = name
-            cell?.characterImage.sd_setImage(with: URL(string: image))
-        }
-        return cell!
+extension CharactersViewController {
+    private func configureDataSource() {
+        dataSource = DataSource(collectionView: charactersGridView.collectionView, cellProvider: { (collectionView, indexPath, character) -> UICollectionViewCell? in
+            return collectionView.dequeueConfiguredReusableCell(using: self.charactersGridView.characterCell, for: indexPath, item: character)
+        })
     }
 }
 
 // MARK: - UICollectionViewDelegate
 extension CharactersViewController: UICollectionViewDelegate {
+
     func collectionView(_ collectionView: UICollectionView, willDisplay cell: UICollectionViewCell, forItemAt indexPath: IndexPath) {
         if indexPath.row == collectionView.numberOfItems(inSection: indexPath.section) - 1 {
+            // show lazy-loading indicator
             charactersGridView.loadingIndicator.startAnimating()
             DispatchQueue.main.asyncAfter(deadline: .now() + 1) {
                 self.loadMore()
                 self.charactersGridView.loadingIndicator.stopAnimating()
             }
```

Comment:
Instead of waiting for 1 second, how about putting the `stopAnimation()` call inside our `sink(receiveValue:)` closure so that the indicator won't disappear too early with slow network connection?